cmd/deploy/input_test.go

```diff
@@ -2,9 +2,234 @@ package main
+			"failed to parse ref:",
+		},
+		{
+			"invalid reference url",
+			typeManifest,
+			[]string{
+				"localhost:5000/foo?test",
+			},
+			want{},
+			"failed to parse ref:",
+		},
+		{
+			"mismatch digest",
+			typeManifest,
+			[]string{
+				"localhost:5000/foo@sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				"localhost:5000/foo@sha256:00042f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			want{},
+			"refs digest mismatch in ",
+		},
+		{
+			"can't push to blob",
+			typeBlob,
+			[]string{
+				"localhost:5000/foo:tag",
+			},
+			want{},
+			"cannot push blobs to a tag:",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			refs, refsDigest, err := normalizeInputRefs(x.deployType, x.rawDigests)
+
+			if x.wantErr != "" {
+				if err == nil {
+					t.Fatalf("Expected error not returned: %s", x.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), x.wantErr) {
+					t.Fatalf("Expected error doesn't match. got:\n'%s',\n\nexpected to contain:\n'%s'", err.Error(), x.wantErr)
+				}
+			}
+
+			if refsDigest != x.want.digest {
+				t.Errorf("got:\n%s\n\nexpected:\n%s\n", string(refsDigest), x.want.digest)
+			}
+
+			if !slices.Equal(refs, x.want.refs) {
+				t.Errorf("Reference doesn't match. got: \n%#v,\n\nexpected:\n%#v", refs, x.want.refs)
+			}
+		})
+	}
+}
+
+func TestNormalizeInputLookup(t *testing.T) {
+	type want struct {
+		lookups map[ociregistry.Digest]registry.Reference
+		digest  ociregistry.Digest
+	}
+	type fixture struct {
+		name       string
+		rawLookups map[string]string
+		want       want
+		wantErr    string
+	}
+
+	for _, x := range []fixture{
+		{
+			"happy path",
+			map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo",
+				"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": "foo/bar:baz",
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": "localhost:5000/example:bar",
+			},
+			want{
+				map[ociregistry.Digest]registry.Reference{
+					"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": {
+						Host:       "docker.io",
+						Repository: "library/example",
+						Tag:        "foo",
+						Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+					},
+					"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": {
+						Host:       "docker.io",
+						Repository: "foo/bar",
+						Tag:        "baz",
+						Digest:     "sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86",
+					},
+					"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": {
+						Host:       "localhost:5000",
+						Repository: "example",
+						Tag:        "bar",
+						Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+					},
+				},
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
```

Comment:
I think this might actually be a bug in the implementation -- maps are not ordered (in fact, Go actively randomizes them, which is why this test fails randomly), but in this case the function cannot have determined which value is the "target" or "common" digest, so it should be returning empty :thinking:

(catching this is a nice win from splitting, I think! :grinning:)

Reply:
```go
	var commonDigest ociregistry.Digest // if we store this out here, we can abuse it later to get the "last" lookup digest (for getting the single key in the case of len(lookup) == 1 without a new loop)
```

aha -- so we should update the function to explicitly clear this value before returning it unless the length of our final return map is exactly 1

then we can _also_ update the code calling this function to use whether or not this value is empty instead of `len(lookup) == 1` to determine whether or not to use it :rocket:

Reply:
Oh wait, empty might mean fallback reference, so more of that later conditional needs to move too :thinking:

Reply:
This actually complicates it quite a bit more.  This function signature might need more thought. :thinking:

(Or we adjust the test, but it feels wrong to have the function return data that's not valid to use _and_ it feels wrong to copy so much conditional logic into the test.)

Reply:
Ok, maybe something like this: (the new test needs updates for both this and the other thread, but at least this fixes the implementation)

<details>
<summary>Patch:</summary>

```patch
diff --git a/cmd/deploy/input.go b/cmd/deploy/input.go
index 2a2a736..1477e77 100644
--- a/cmd/deploy/input.go
+++ b/cmd/deploy/input.go
@@ -76,43 +76,48 @@ func normalizeInputRefs(deployType deployType, rawRefs []string) ([]registry.Ref
 	return refs, commonDigest, nil
 }
 
-func normalizeInputLookup(rawLookup map[string]string) (map[ociregistry.Digest]registry.Reference, ociregistry.Digest, error) {
-	lookups := make(map[ociregistry.Digest]registry.Reference, len(rawLookup))
-	var commonDigest ociregistry.Digest // if we store this out here, we can abuse it later to get the "last" lookup digest (for getting the single key in the case of len(lookup) == 1 without a new loop)
+func normalizeInputLookup(rawLookup map[string]string) (map[ociregistry.Digest]registry.Reference, *ociregistry.Digest, error) {
+	lookup := make(map[ociregistry.Digest]registry.Reference, len(rawLookup))
+	var digest ociregistry.Digest // if we store this out here, we can abuse it later to get the "last" lookup digest (for getting the single key in the case of len(lookup) == 1 without a new loop)
 	for d, refString := range rawLookup {
-		commonDigest = ociregistry.Digest(d)
-		if commonDigest != "" {
+		digest = ociregistry.Digest(d)
+		if digest != "" {
 			// normal.Lookup[""] is a special case for fallback (where to look for any child object that isn't explicitly referenced)
-			if err := commonDigest.Validate(); err != nil {
-				return nil, "", fmt.Errorf("lookup key %q invalid: %w", commonDigest, err)
+			if err := digest.Validate(); err != nil {
+				return nil, nil, fmt.Errorf("lookup key %q invalid: %w", digest, err)
 			}
 		}
 		if ref, err := registry.ParseRef(refString); err != nil {
-			return nil, "", fmt.Errorf("failed to parse lookup ref %q: %v", refString, err)
+			return nil, nil, fmt.Errorf("failed to parse lookup ref %q: %v", refString, err)
 		} else {
-			if ref.Tag != "" && commonDigest != "" {
+			if ref.Tag != "" && digest != "" {
 				//return normal, fmt.Errorf("%s: tag on by-digest lookup ref makes no sense: %s (%s)", debugId, ref, d)
 			}
 
-			if ref.Digest == "" && commonDigest != "" {
-				ref.Digest = commonDigest
+			if ref.Digest == "" && digest != "" {
+				ref.Digest = digest
 			}
-			if commonDigest != "" && ref.Digest != commonDigest {
-				return nil, "", fmt.Errorf("digest on lookup ref should either be omitted or match key: %s vs %s", ref, d)
+			if digest != "" && ref.Digest != digest {
+				return nil, nil, fmt.Errorf("digest on lookup ref should either be omitted or match key: %s vs %s", ref, d)
 			}
 
-			lookups[commonDigest] = ref
+			lookup[digest] = ref
 		}
 	}
 
-	return lookups, commonDigest, nil
+	// see notes on "digest" definition
+	if len(lookup) != 1 {
+		return lookup, nil, nil
+	}
+	if digest == "" && (lookup[""].Digest == "" && lookup[""].Tag == "") {
+		// if it was a fallback, it needs at least Tag or Digest (or our refs need Digest, so we can infer)
+		return lookup, nil, nil
+	}
+	return lookup, &digest, nil
 }
 
 func NormalizeInput(raw inputRaw) (inputNormalized, error) {
 	var normal inputNormalized
-	var lookupDigest ociregistry.Digest
-	var refsDigest ociregistry.Digest
-	var err error
 
 	switch raw.Type {
 	case "":
@@ -132,12 +137,18 @@ func NormalizeInput(raw inputRaw) (inputNormalized, error) {
 	if len(raw.Refs) == 0 {
 		return normal, fmt.Errorf("zero refs specified for pushing (need at least one)")
 	}
+	var (
+		refsDigest ociregistry.Digest
+		err error
+	)
 	normal.Refs, refsDigest, err = normalizeInputRefs(normal.Type, raw.Refs)
 	if err != nil {
 		return normal, err
 	}
 
-	debugId := normal.Refs[0]
+	debugId := normal.Refs[0] // used for annotating errors from here out
+
+	var lookupDigest *ociregistry.Digest
 	normal.Lookup, lookupDigest, err = normalizeInputLookup(raw.Lookup)
 	if err != nil {
 		return normal, fmt.Errorf("%s: %w", debugId, err)
@@ -147,29 +158,24 @@ func NormalizeInput(raw inputRaw) (inputNormalized, error) {
 		// if we have no Data, let's see if we have enough information to infer an object to copy
 		if lookupRef, ok := normal.Lookup[refsDigest]; refsDigest != "" && ok {
 			// if any of our Refs had a digest, *and* we have a way to Lookup that digest, that's the one
-			lookupDigest = refsDigest
+			lookupDigest = &refsDigest
 			normal.CopyFrom = &lookupRef
-		} else if lookupRef, ok := normal.Lookup[lookupDigest]; len(normal.Lookup) == 1 && ok {
+		} else if lookupDigest != nil {
 			// if we only had one Lookup entry, that's the one
-			if lookupDigest == "" {
-				// if it was a fallback, it needs at least Tag or Digest (or our refs need Digest, so we can infer)
-				if lookupRef.Digest == "" && lookupRef.Tag == "" {
-					if refsDigest != "" {
-						lookupRef.Digest = refsDigest
-					} else {
-						return normal, fmt.Errorf("%s: (single) fallback needs digest or tag: %s", debugId, lookupRef)
-					}
-				}
-			}
+			lookupRef := normal.Lookup[*lookupDigest]
+			normal.CopyFrom = &lookupRef
+		} else if lookupRef, ok := normal.Lookup[""]; refsDigest != "" && ok {
+			lookupDigest = &refsDigest
+			lookupRef.Digest = refsDigest
 			normal.CopyFrom = &lookupRef
 		} else {
 			// if Lookup has only a single entry, that's the one (but that's our last chance for inferring intent)
-			return normal, fmt.Errorf("%s: missing data (and lookup is not a single item)", debugId)
+			return normal, fmt.Errorf("%s: missing data (and lookup is not a single item or fallback with digest or tag)", debugId)
 			// TODO *technically* it would be fair to have lookup have two items if one of them is the fallback reference, but it doesn't really make much sense to copy an object from one namespace, but to get all its children from somewhere else
 		}
 
-		if lookupDigest == "" && normal.CopyFrom.Digest != "" {
-			lookupDigest = normal.CopyFrom.Digest
+		if *lookupDigest == "" && normal.CopyFrom.Digest != "" {
+			lookupDigest = &normal.CopyFrom.Digest
 		}
 
 		if _, ok := normal.Lookup[""]; !ok {
@@ -178,8 +184,8 @@ func NormalizeInput(raw inputRaw) (inputNormalized, error) {
 		}
 
 		if refsDigest == "" {
-			refsDigest = lookupDigest
-		} else if lookupDigest != "" && refsDigest != lookupDigest {
+			refsDigest = *lookupDigest
+		} else if *lookupDigest != "" && refsDigest != *lookupDigest {
 			return normal, fmt.Errorf("%s: copy-by-digest mismatch: %s vs %s", debugId, refsDigest, normal.CopyFrom)
 		}
 	} else {
```

</details>